### PR TITLE
Refactor Codebase, Update CI, and Restructure Data Handling

### LINKED ISSUE #2: Potential refactor for `LyfcalApp::initialize`
The current implementation of `initialize` looks like this:

https://github.com/Mutualideas/lyfcal/blob/f8a286628b4718e2d9bd563e047a10320d83d736/src/main.rs#L89-L104

This is nice and clean.

However, it's worth noting that while we take `&mut self` as an argument, only the `draw_data` field is actually mutated.

## `DrawData::initialize`

One change that might help to make the code a bit more modular is to move the `initalize` method to the `DrawData` type. E.g.

```rust
impl DrawData {
    fn initialize(&mut self, config: &Config) {
        self.birthdate = config.birthdate; 
        self.elapsed_date_bool = config.elapsed_date_bool;
        // ...
        // etc
    }
}
```

Then, our original `LyfcalApp::initialize` method can become:
```rust
// When initialized, data from the config is passed to draw_data. 
fn initialize(&mut self) {
    self.draw_data.initialize(&self.config);
}
```

## Adding a `Config` field to `DrawData`

Another thing worth noting is that `DrawData` seems to contain an exact copy of every field in `Config`. We can consider having `DrawData` simply have its own `Config` field to make this initialize step much simpler, e.g.

```rust
#[derive(Debug)]
struct DrawData {
    /// A snapshot of the main application `Config`.
    conifg: Config,
    events: BTreeMap<NaiveDate, String>,
}
```

Then, rather than having an `initialize` method for `DrawData`, we can just clone the main `Config`, e.g. in `LyfcalApp::initialize`:
```rust
// When initialized, data from the config is passed to draw_data. 
fn initialize(&mut self) {
    self.draw_data.config = self.config.clone(); // Note: this may require `#[derive(Clone)]` on our `struct Config` declaration.
}
```

### PR DESCRIPTION
**Pull request by mutualideas**
This pull request includes several key updates and improvements to the project:

1. Codebase Cleanup:

- General code cleanup to improve readability and maintainability.

2. Merge Latest Changes from Main:
- Merged the latest changes from the main branch to ensure the codebase is up-to-date.

3. Main.rs Update:
- Made updates to the main.rs file to enhance functionality and address existing issues.

4. Continuous Integration (CI) Improvements:

- Amended the rust.yml CI configuration file to improve the continuous integration process.
- Fixed a type issue in the CI configuration.

5. Function Refactoring and Data Restructuring:

- Refactored the update function to optimize performance and maintainability.
- Restructured the data handling within the draw function to improve efficiency and clarity.

These changes collectively enhance the project's code quality, ensure CI processes are robust, and improve the overall functionality and structure of the application.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -10,27 +10,21 @@ env:
...
-        override: true
 
-    
-    - name: Build
-      run: cargo build --verbose
+    - uses: actions/checkout@v3
 
-    - name: Run tests
-      run: cargo test --locked --verbose
+    - uses: actions-rs/toolchain@v1
```
Comment (mitchmindtree) [inline .github/workflows/rust.yml:26]: Hmmmmm looks like it's failing on this action.

The error says no `rust-toolchain` file is present in the repo, but this shouldn't be needed.

What if you change this line too:

```suggestion
    - uses: actions-rs/toolchain@v1
      with:
        toolchain: stable
```

This just specifies the `stable` toolchain, though tbh I thought this was the default :thinking:

**Comment by mitchmindtree** [comment]
Nice - now you can start addressing [the clippy warnings](https://github.com/Mutualideas/lyfcal/actions/runs/9593475140/job/26453987624?pr=5#step:5:470) :rocket: You can run clippy locally with `cargo clippy` to get the same warnings locally.

**Comment by mutualideas** [comment]
I'm addressing them now. I assume there is a way to append the changes to an existing pull request. Clippy is such a hardass—I love it!

**Comment by mitchmindtree** [comment]
> I assume there is a way to append the changes to an existing pull request

Yes, each time you push a commit to this branch it will automatically show up in the pull request :+1:

**Pull request merged**